File: nexxT/tests/src/TestExceptionFilter.cpp
/* 
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (C) 2020 ifm electronic gmbh
 *
 * THE PROGRAM IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
 */

#include "TestExceptionFilter.hpp"
#include "PropertyCollection.hpp"

TestExceptionFilter::TestExceptionFilter(BaseFilterEnvironment *env)
    : Filter(false, false, env)
{
    propertyCollection()->defineProperty("whereToThrow", "nowhere", "one of nowhere,constructor,init,start,port,stop,deinit");
    if( propertyCollection()->getProperty("whereToThrow") == "constructor" )
    {
        throw std::runtime_error("exception in constructor");
    }
    port = SharedInputPortPtr(new InputPortInterface(false, "port", env, 1, -1.));
    addStaticPort(port);
}

TestExceptionFilter::~TestExceptionFilter()
{
    /*
     * c++11 has the convention to terminate() when destructors throw exceptions
     * we stick to this and do not support throwing exceptions in filter destructors.
     */
}


void TestExceptionFilter::onInit()
{
    if( propertyCollection()->getProperty("whereToThrow") == "init" )
    {
        throw std::runtime_error("exception in init");
    }
}

void TestExceptionFilter::onOpen()
{
    if( propertyCollection()->getProperty("whereToThrow") == "open" )
    {
        throw std::runtime_error("exception in open");
    }
}

void TestExceptionFilter::onStart()
{
    if( propertyCollection()->getProperty("whereToThrow") == "start" )
    {
        throw std::runtime_error("exception in start");
    }
}

void TestExceptionFilter::onPortDataChanged(const InputPortInterface &)
{
    if( propertyCollection()->getProperty("whereToThrow") == "port" )
    {
        throw std::runtime_error("exception in port");
    }
}

void TestExceptionFilter::onStop()
{
    if( propertyCollection()->getProperty("whereToThrow") == "stop" )
    {
        throw std::runtime_error("exception in stop");
    }
}

void TestExceptionFilter::onClose()
{
    if( propertyCollection()->getProperty("whereToThrow") == "close" )
    {
        throw std::runtime_error("exception in close");
    }
}

void TestExceptionFilter::onDeinit()
{
    if( propertyCollection()->getProperty("whereToThrow") == "deinit" )
    {
        throw std::runtime_error("exception in deinit");
    }
}

File: nexxT/tests/interface/TestExceptionFilter.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (C) 2020 ifm electronic gmbh
#
# THE PROGRAM IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
#

from nexxT.interface import Filter, InputPort

class TestExceptionFilter(Filter):
    def __init__(self, env):
        super().__init__(False, False, env)
        self.propertyCollection().defineProperty("whereToThrow", "nowhere",
                                                 "one of nowhere,constructor,init,open,start,port,stop,close,deinit")
        if self.propertyCollection().getProperty("whereToThrow") == "constructor":
            raise RuntimeError("exception in constructor")
        self.port = InputPort(False, "port", env)
        self.addStaticPort(self.port)

    def onInit(self):
        if self.propertyCollection().getProperty("whereToThrow") == "init":
            raise RuntimeError("exception in init")

    def onOpen(self):
        if self.propertyCollection().getProperty("whereToThrow") == "open":
            raise RuntimeError("exception in open")

    def onStart(self):
        if self.propertyCollection().getProperty("whereToThrow") == "start":
            raise RuntimeError("exception in start")

    def onStop(self):
        if self.propertyCollection().getProperty("whereToThrow") == "stop":
            raise RuntimeError("exception in stop")

    def onClose(self):
        if self.propertyCollection().getProperty("whereToThrow") == "close":
            raise RuntimeError("exception in close")

    def onDeinit(self):
        if self.propertyCollection().getProperty("whereToThrow") == "deinit":
            raise RuntimeError("exception in deinit")

    def onPortDataChanged(self, port):
        if self.propertyCollection().getProperty("whereToThrow") == "port":
            raise RuntimeError("exception in port")
